runtime/lua/vim/fswatch/init.lua

```diff
@@ -0,0 +1,123 @@
+--[[
+  Create a file watcher, each watcher is identified by the name of the file that it
+  watches. We can use a lua table to store all watchers indexed by their filenames
+  so that we can close the required watcher during the callback to on_change to
+  debounce the watcher.
+--]]
+
+local uv = vim.loop
+
+local Watcher = {}
+local WatcherList = {}
+
+-- idle handle to check for any pending notifications in a watcher.
+-- Only displays the notifications if neovim is in focus, and the buffer
+-- is the current buffer.
+
+-- Callback for the idle handle
+function check_notifications()
+  for f, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs and watcher.paused == false then
+      if uv.fs_stat(watcher.ffname) ~= nil then
+        vim.api.nvim_command('call fswatch#PromptReload()')
+      else
+        print("ERR: File "..watcher.fname.." removed")
+      end
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+local check_handle = uv.new_check()
+check_handle:start(vim.schedule_wrap(check_notifications))
+
+function Watcher:new(fname)
+  assert(fname ~= '', 'Watcher.new: Error: fname is an empty string')
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {fname = fname, ffname = ffname, handle = nil, paused = false, pending_notifs = false}
+  setmetatable(w, self)
+  self.__index = self
+  return w
+end
+
+function Watcher:start()
+  assert(self.fname ~= '', 'Watcher.start: Error: no file to watch')
+  assert(self.ffname ~= '', 'Watcher.start: Error: full path for file not available')
+  -- get a new handle
+  self.handle = uv.new_fs_event()
+  self.handle:start(self.ffname, {}, self.on_change)
+end
+
+function Watcher:stop()
+  assert(self.fname ~= '', 'Watcher.stop: Error: no file being watched')
+  assert(self.handle ~= nil, 'Watcher.stop: Error: no handle watching the file')
+  self.handle:stop()
+  -- close the handle altogether, for windows.
+  if self.handle:is_closing() then
+    return
+  end
+  self.handle:close()
+end
+
+function Watcher.on_change(err, fname, event)
+  WatcherList[fname].pending_notifs = true
+
+  WatcherList[fname]:stop()
+  WatcherList[fname]:start()
+end
+
+function Watcher.start_watch(fname)
+  -- since we can only get file name from callback, use only the file
+  -- name for storing in table. (Without the rest of the path.)
+  local f = vim.api.nvim_call_function('fnamemodify', {fname, ':t'})
```

Comment:
Wait so will this not work for two files with the same tail name? That doesn't seem good.

Reply:
I was thinking of using `fnamemodify` to get full path in the `on_change` callback, but then `vim.api.nvim_call_function` causes some problems inside the callback too so couldn't decide anything. I can try that for now, but do you think we can try something else? 

I thought of using buffer numbers instead of the filenames, but it too would mean calling `nvim_call_function` inside the callback to get the buffer number from file name.

Reply:
calling functions from lua is very fast. do not worry about calling them. Just make sure it's does the right stuff :) We can fix perf afterwards.

Reply:
@tjdevries, should not be a problem now :)